app/models/concerns/text_sanitizer.rb

```diff
@@ -37,4 +42,21 @@ def sanitize_category(kw, max_length, strict)
       kw.strip.slice(0, max_length)
     end
   end
+
+  def add_newlines_to_tags(text)
+    text.gsub(/<p>/i, "\n<p>")
+      .gsub(/<\/p>/i, "</p>\n")
+      .gsub(/<div>/i, "\n<div>")
+      .gsub(/<\/div>/i, "</div>\n")
+      .gsub(/<br>/i, "\n<br>")
+      .gsub(/<br\s*\/>/i, "\n<br>")
+  end
+
+  def clean_whitespace(text)
+    text.gsub(/\R/, "\n") # if there is any kind of newline, make it \n
+      .gsub(/([ \t]*)\n([ \t]*)/, "\n") # get rid of whitespace on eother side of a newline
+      .gsub(/\n\n\n/, "\n\n") # no more than 2 newlines together
```

Comment:
Ah, maybe this covers my previous questions... even if we had too many newlines, this should remove some of them.

Are these rules from some real-world cases you've seen, or just from the test?

I also wonder if the last steps (trimming/combining spaces) should happen before the multiple-newline detection.  But this is probably fine.

Reply:
the `strip` could be before or after I guess, doesn't seem to matter much.
the prior regex gets rid of spaces near newlines the final squeeze is any place with multiple spaces, so again different, and could be any order

Reply:
+1